File: zZ/cpp/longest-substring-with-at-most-two-distinct-characters.cpp
// 159. Longest Substring with At Most two Distinct Characters
// Difficulty : Hard 

// Given a string, find the length of the longest substring T that contains at most 2 distinct characters.

// For example, Given s = “eceba” and k = 2,

// T is "ece" which its length is 3.

// Time:  O(n)
// Space: O(1)

class Solution {
public:
  int lengthOfLongestSubstringTwoDistinct(string s) {
    const int k = 2;
    int longest = 0, start = 0, distinct_count = 0;
    array<int, 256> visited = {0};
    for (int i = 0; i < s.length(); ++i) {
      if (visited[s[i]]++ == 0) {
        ++distinct_count;
      }
      while (distinct_count > k) {
        if (--visited[s[start]] == 0) {
          --distinct_count;
        }
        ++start;
      }
      longest = max(longest, i - start + 1);
    }
    return longest;
  }
};

File: zZ/java/longest-substring-with-at-most-two-distinct-characters.java
/**
 * Longest Substring with At Most Two Distinct Characters
 * 
 * Given a string S, find the length of the longest substring T that contains at most two distinct characters.
 * For example,
 * Given S = “eceba”,
 * T is “ece” which its length is 3.
 */

import java.util.HashMap;
import java.util.Map;

public class Solution {
    public int lengthOfLongestSubstringTwoDistinct(String s) {
        Map<Character, Integer> characterCount = new HashMap<>();

        int maxLength = 0;
        int start = 0;
        int distinctCount = 2;
        for (int i = 0; i < s.length(); ++i) {
            characterCount.put(s.charAt(i), characterCount.getOrDefault(s.charAt(i), 0) + 1);

            while (characterCount.size() > distinctCount) {
                characterCount.put(s.charAt(start), characterCount.get(s.charAt(start)) - 1);
                if (characterCount.get(s.charAt(start)) == 0) {
                    characterCount.remove(s.charAt(start));
                }
                ++start;
            }

            maxLength = Math.max(maxLength, i - start + 1);
        }

        return maxLength;
    }
}
